### DISCUSSION TITLE
Synchronize video to match prebuffered audio

### DISCUSSION THREAD

**Discussion by asbjornu** [dosbox-staging/dosbox-staging#2243] [Q&A]
I'm in the process of capturing all of my old demoscene tracker modules with DOSBox Staging and have found that some modules require a fairly high `blocksize` and/or `prebuffer` to not stutter. I don't mind adding several seconds of delay to the audio to avoid stuttering, but I do mind the video and audio becoming unsynchronized.

Is it possible to buffer or add some sort of delay to the video so it becomes synchronized with a fairly delayed audio stream, somehow?

**Comment by kcgen** [comment]
Can you post how to reproduce this? (Player, module(s), conf settings). 
Ideally a zip with them all pre-setup 👍

**Reply by asbjornu** [reply]
I fiddled a bit with the settings now and found that turning off `vsync` helped dramatically. I would prefer to have `vscyn` on to avoid tearing, but I haven't noticed any tearing yet, so perhaps I can keep it off. I'll post a reproduction if turning off `vsync` didn't completely solve the problem or if I decide that I need `vsync` on for whatever reason.

**Comment by kcgen** [comment]
Default setting can be tested with `-noprimaryconf -nolocalconf` (to block it from reading conf files).

 When defaults are used, vsync will only be turned on if it detects that vsync is being enforced on the host-side (and frame rates are throttled/blocked to once every 16ms or so). 

This doesn't affect the DOS side; vsync cannot be forced onto the DOS application, which will either be programmed to draw on vblank intervals or it won't be... And it will tear (like Heretic and System Shock). 

Good luck. Please share your results 🙂

**Reply by asbjornu** [reply]
Thanks, I now tried to play a problematic song with `-noprimaryconf -nolocalconf` and there is definitely stuttering going on there. 

I'm trying to play [this S3M file](https://api.modarchive.org/downloads.php?moduleid=199723#poc_jazz.s3m) in [ScreamTracker 3](https://files.scene.org/view/resources/music/trackers/scrmt321.zip). The stuttering becomes very apparent in pattern `02` (about 23 seconds into the song). 

To load and play the song in ScreamTracker 3, start `ST3.EXE`, press CTRL+L, locate the song, press enter to load it, then F5 to play it. If you are able to reproduce this, that would be fantastic!

**Reply by asbjornu** [reply]
As a note, increasing `blocksize` to `8192`, `prebuffer` to `500` and `cycles` to more than `10000` seems to alleviate the stuttering. What's interesting is that keeping these values low (`1024`, `100` and `auto`, respectively) works fine if I set `gus` to `true` and play through Gravis Ultrasound instead of SoundBlaster 16. Weird, no?

**Comment by kcgen** [comment]
Running ScreamTracker 3 (developed in ~1993/1994) with `cycles = 20000` feels about right, and doesn't stutter (atleast for me).

Some general tips (applicable to most DOSBoxes) - If you're using a modern Linux distro, you can ask SDL to send audio via pipewire  (and X11) with these environment variables:
```
export SDL_AUDIODRIVER=pipewire
export SDL_VIDEODRIVER=x11
```

Also if you're on Linux, try to eliminate all competing screen compositing load (minimize terminals, browsers, etc.)

Use an external recording tool (like OBS) with GPU-based encoding to minimize the impact to DOSBox.

If you use DOSBox's built-in recorder (Ctrl+F7), be aware that it's not multi-threaded. The passage of time at the host level can be impacted during high-stress events (like the mod player doing a DMA transfer to the audio device), however, the passage of time inside DOS-land will be correct and the recorded audio will not have those audio gaps.  So you can still use Ctrl+F7 if you like.

I can also highly recommend using Impulse Tracker 2.14 B (with `cycles = 30000`), which has a very similar interface to ST3. @8bitbubsy (https://16-bits.org/index.php) might also have better recommendations for DOS when it comes to MOD/S3M/XM/IT players.

Here's a recording from IT 2.14 B using the current build of DOSBox Staging, using `Ctrl+F7` recording:


https://user-images.githubusercontent.com/1557255/214190509-cece158d-fc1c-4944-9603-fdea4a016a83.mp4

**Reply by asbjornu** [reply]
Thank you so much! Setting `cycles` to `20000` indeed removed the stuttering. I was not aware of the <kbd>CTRL</kbd><kbd>F7</kbd> shortcut; I have been screen recording with QuickTime (I'm on a Mac). However, neither <kbd>CTRL</kbd><kbd>F7</kbd> nor <kbd>⌘</kbd><kbd>F7</kbd> seem to do anything for me? Or, perhaps DOSBox Staging is recording, but there's no visual feedback after pressing it and I don't know where the recorded file is stored. Hints would be welcome. 😅 

Regarding Impulse Tracker, I love that thing to bits as I have produced more than a 100 `.IT` files with it. 😃 However, I do have some `.MOD`, `.S3M` and `.XM` files as well, and just for variety and faithfulness, I'm trying to capture every file with an approximation of the same tracker in which it was originally produced. So that means using @8bitbubsy's fantastic _ProTracker 2 clone_ for `.MOD` files and his wonderful _FastTracker II clone_ for `.XM` files. For `.IT` files I haven't made up my mind yet whether to use Impulse Tracker within DOSBox Staging or [Schism Tracker](http://schismtracker.org/). For `.S3M` files, there are no native Scream Tracker 3 clones afaik., so I have to use DOSBox Staging, hence this Q/A. :)

Thank you so much for the assistance, @kcgen! 🙏🏼  Highly appreciated! ❤️

**Reply by asbjornu** [reply]
Okay, I figured out the capturing by configuring `captures` directory in `dosbox-staging.conf` and pressing <kbd>⌘</kbd><kbd>F7</kbd> to start/stop capturing.

It looks like audio is captured in lossless PCM, which is excellent. However, the video is captured in 640 x 400 pixels, which is too low. I need the video to be at least in full HD to get the best sound and video quality from YouTube. Can the resolution of the video capture be increased, somehow?

It's of course possible to upscale the video with ffmpeg or similar afterward, but it would save time and yield a better result to just capture the fullscreen resolution initially.

**Comment by johnnovak** [comment]
Some thoughts about this:

In general, you can delay the video by the audio playback latency to get perfect sync; I think some Windows era demos actually do something of the sort.

Of course, that only works for non-interactive stuff when you're just watching a demo. For interactive programs, you're increasing your input latency, so you'll have to wait say 50-100ms for keystrokes to appear on the screen when typing. Not good 😄 

To make trackers really useful under DOSBox (to achieve the same near-zero latency audio experience you have under DOS), we'd need ASIO driver support; that's pretty much the only way to achieve very low latency audio on macOS and Windows (but if you're a hobby musician, you probably already know that). Of course, you need "prosumer" level audio interfaces for that, but there's no other way. For that we'd need to intoduce the RtAudio library into the project for audio support, that would give us the ability to use ASIO drivers in a cross-platform manner. But it's a non-trivial amount of work, so maybe later... (I'm personally interested in this, but have too many things on my plate already...)

About capturing larger than 640x400, ideally we'd have the ability to capture the post-shader output, then you could just use any shader, or the default "sharp" shader to scale the output to any resolution you like.

In practice, for now I think you'll just need to do the video scaling post-capture as a post-processing step. Same goes for syncing with the audio — just use the same buffer size consistently for your captures, then adjust for the latency in post-processing. Well, at least that's how I'd approach it if I wanted to make some videos with perfect sync today 😄 

Hope that helps!

**Reply by asbjornu** [reply]
> Of course, that only works for non-interactive stuff when you're just watching a demo. For interactive programs, you're increasing your input latency, so you'll have to wait say 50-100ms for keystrokes to appear on the screen when typing.

For a screen recording of Scream Tracker 3 playing a song, the input delay could be a minute for what I care. 😄

> To make trackers really useful under DOSBox (to achieve the same near-zero latency audio experience you have under DOS), we'd need ASIO driver support;

ASIO is Windows-only, no?

> About capturing larger than 640x400, ideally we'd have the ability to capture the post-shader output, then you could just use any shader, or the default "sharp" shader to scale the output to any resolution you like.

That would be great!

> In practice, for now I think you'll just need to do the video scaling post-capture as a post-processing step.

I've tried that and found I get better video results by capturing the full (post-shader) resolution with QuickTime. That, however, results in a lossy audio stream, which is sub-optimal. So I'm now experimenting with [OBS](https://obsproject.com/) to have the video captured in h.264, while capturing the audio in lossless FLAC, all to an MKV container. It requires a lot of resources during capturing, but it might just work.

**Reply by johnnovak** [reply]
> ASIO is Windows-only, no?

Yeah, but the point of PortAudio is that is supports the lowest-latency driver formats per each platform. So ASIO on Windows, CoreAudio on macOS, and whatever people use on Linux these days (I wouldn't now, not a Linux user).

> So I'm now experimenting with [OBS](https://obsproject.com/) to have the video captured in h.264, while capturing the audio in lossless FLAC, all to an MKV container. It requires a lot of resources during capturing, but it might just work.

I think something like that is your best best for now (unless you can apply some different CRT shader in post-processing).

**Reply by asbjornu** [reply]
> > ASIO is Windows-only, no?
> 
> Yeah, but the point of PortAudio is that is supports the lowest-latency driver formats per each platform. So ASIO on Windows, CoreAudio on macOS, and whatever people use on Linux these days (I wouldn't now, not a Linux user).

Yes, that makes sense.

> > So I'm now experimenting with [OBS](https://obsproject.com/) to have the video captured in h.264, while capturing the audio in lossless FLAC, all to an MKV container. It requires a lot of resources during capturing, but it might just work.
> 
> I think something like that is your best best for now (unless you can apply some different CRT shader in post-processing).

I've explored OBS quite a bit now and unfortunately have to conclude that it too doesn't quite work. I've now tried the following options, neither of which give satisfactory results (yet):

### Low-resolution capture with OBS

I can run DOSBox Staging in low resolution, capture with OBS, and upscale in post-processing. However, since I'm unable to tweak OBS to capture pixel perfect (even in the "Lossless Quality, Tremendously Large File Size" recording quality), anti-aliasing is applied which is multiplied and massively worsened when upscaling.

### Low-resolution capture with DOSBox Staging

I can run DOSBox Staging in low resolution, capture with <kbd>⌘</kbd><kbd>F7</kbd>, and upscale in post-processing. Since DOSBox Staging captures the video pixel-perfect (no anti-aliasing), upscaling with the `neighbor` algorithm works well. This would have been the best option, had it not been for sound stuttering no matter what I set `blocksize` and `prebuffer` to (currently at `8192` and `2000`).

### Fullscreen resolution capture with OBS

I can run DOSBox Staging in fullscreen and capture with OBS. This, however, leads to sound stuttering since capturing anything above 720p with OBS on a Mac requires so much CPU that there's nothing left for DOSBox Staging. OBS requires so much CPU because it doesn't support hardware acceleration on Macs with AMD GPUs.

(I have an iMac (Retina 5K, 27-inch, 2020) with 3,6 GHz 10-Core Intel Core i9, 32 GB 2667 MHz DDR4, and AMD Radeon Pro 5700 XT 16 GB. I think OBS is at fault for not being able to utilize this hardware better than it currently is. But I don't think there's much I can do about it.)

### Fullscreen resolution capture with QuickTime

I can run DOSBox Staging in fullscreen and capture with QuickTime. Since QuickTime is able to take advantage of the GPU, this works without sound stuttering, but for some inexplicable reason, QuickTime captures sound in lossy AAC with a rather low bitrate. As the point of the capture is to upload the video to YouTube and other video hosting services, the sound will be re-encoded from the low bitrate AAC, causing a lot of damage to the sound.

If you have ideas for how to tweak DOSBox Staging's options or try anything I haven't already, I'm all ears. Thank you so much for the time you're spending helping me!

**Reply by GiulianoPastorino** [reply]
Hi @asbjornu,

Regarding prebuffer, you mentioned that:

> This would have been the best option, had it not been for sound stuttering no matter what I set blocksize and prebuffer to (currently at 8192 and 2000).

The other day I tried a prebuffer of 1000, and the log showed me the following message:

`CONFIG: Invalid 'prebuffer' setting: '1000'. Value outside of the valid range 0-100, using '100'`

So, is it possible to use a prebuffer above 100?

Regarding stuttering, I believe this is a sound issue that is unrelated to "pops" or "clicks" in the sound, am I correct?

Cheers.

**Reply by johnnovak** [reply]
> So, is it possible to use a prebuffer above 100?

No, and it would be rather pointless because of the huge lag.

**Reply by asbjornu** [reply]
> The other day I tried a prebuffer of 1000, and the log showed me the following message:
> 
> `CONFIG: Invalid 'prebuffer' setting: '1000'. Value outside of the valid range 0-100, using '100'`

@GiulianoPastorino, ah, I haven't noticed that log message.

> Regarding stuttering, I believe this is a sound issue that is unrelated to "pops" or "clicks" in the sound, am I correct?

It's now been so long since I last had a look at this, but IIRC, it was indeed "pops" and "clicks" in the sound that was the problem.

> > So, is it possible to use a prebuffer above 100?
> 
> No, and it would be rather pointless because of the huge lag.

@johnnovak, not if the use of the DOSbox isn't for real-time purposes (such as gaming). As noted, my use case is for recording videos of old trackers playing tracker modules. Whether there's no delay or a whole minute delay doesn't matter to me, as long as the capture of video and audio is pristine.

**Reply by johnnovak** [reply]
@asbjornu You've brought up different (but related) issues in this thread, but it's hard to help you unless you provide a concrete repro case with exact reproduction steps.

The upcoming 0.82.0 RC will continue substantial audio improvements. So wait for that, play around with it, then please raise an issue ticket if you're still having audio issues.

Btw, predelay should help with glitches for the realtime audio you hear when running Staging. The captured audio should be pristine even if the realtime sound glitches like crazy. Anyway, very hard to pinpoint what your issue is without you answering all questions in detail in our issue template.
